### PR TITLE
Api

### PR DESCRIPTION
**Pull request by mahamdeh0**

### CODE REVIEW

File: RestaurantReservation.API/Controllers/AuthenticationController.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace RestaurantReservation.API.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class AuthenticationController : ControllerBase
+    {
+        private readonly JwtTokenGenerator _jwtTokenGenerator;
+
+        public AuthenticationController(JwtTokenGenerator jwtTokenGenerator)
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Controllers/AuthenticationController.cs]: you should be injecting the interface here IJwtTokenGenerator

File: RestaurantReservation.API/Controllers/AuthenticationController.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            _jwtTokenGenerator = jwtTokenGenerator;
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Controllers/AuthenticationController.cs]: with .Net 8 constructors can be simplified to a single line
```
public class AuthenticationController(IJwtTokenGenerator jwtTokenGenerator) : ControllerBase
```

File: RestaurantReservation.API/Services/JwtTokenGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            _configuration = configuration;
+        }
+
+        public string GenerateToken(string UserName, string Password)
+        {
+            var claims = new[]
+            {
+               new Claim(ClaimTypes.Name, UserName),
+               new Claim("Password", Password)
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Services/JwtTokenGenerator.cs]: why are you adding password in claims ? won't that be exposing sensitive information

File: RestaurantReservation.API/Controllers/CustomersController.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<ActionResult<IEnumerable<CustomerDto>>> GetAllCustomers(int pageNumber = 1, int pageSize = 10)
+        {
+            if (pageNumber < 1 || pageSize < 1)
+                return BadRequest($"'{nameof(pageNumber)}' and '{nameof(pageSize)}' must be greater than 0.");
+
+            pageSize = Math.Min(pageSize, MaxPageSize);
+
+            var (customers, paginationMetadata) = await _customerRepository.GetAllAsync(_ => true, pageNumber, pageSize);
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Controllers/CustomersController.cs:44]: what do you think about accessing the repo directly in controller vs adding another layer between them ?

File: RestaurantReservation.API/Controllers/CustomersController.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        /// </summary>
+        /// <param name="id">The ID of the customer to retrieve.</param>
+        /// <returns>A customer DTO if found; otherwise, a 404 Not Found response.</returns>
+        [HttpGet("{id}", Name = "GetCustomer")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<ActionResult<CustomerDto>> GetCustomer(int id)
+        {
+            var customer = await _customerRepository.GetByIdAsync(id);
+            if (customer == null)
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Controllers/CustomersController.cs:62]: because you have the nullable annotation enabled, you need to mark the object as a nullable value before checking if it's null. the way it's implemented it will fire a different exception if it was null

File: RestaurantReservation.API/Controllers/EmployeesController.cs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        /// <param name="pageNumber">The number of the page to retrieve.</param>
+        /// <param name="pageSize">The number of items per page.</param>
+        /// <returns>A list of employee DTOs.</returns>
+        [HttpGet]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<ActionResult<IEnumerable<EmployeeDto>>> GetEmployees(int pageNumber = 1, int pageSize = 10)
+        {
+            if (pageNumber < 1 || pageSize < 1)
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Controllers/EmployeesController.cs:43]: this logic so far was repeated twice in controllers, it can be moved to a common class

File: RestaurantReservation.API/Controllers/EmployeesController.cs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        /// Calculates the average order amount handled by a specific employee.
+        /// </summary>
+        /// <param name="employeeId">The ID of the employee for whom to calculate the average order amount.</param>
+        /// <returns>A 200 OK response with the average order amount; or a 404 Not Found response if no orders exist for the specified employee.</returns>
+        [HttpGet("{employeeId}/average-order-amount")]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<IActionResult> GetAverageOrderAmount(int employeeId)
+        {
+            try
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Controllers/EmployeesController.cs]: do you think it's the responsibility of the controller to catch and map exceptions ? and why is it only this endpoint that has this ?

File: RestaurantReservation.API/Models/Customers/CustomerDto.cs
Comment (hiba-kurd) [inline RestaurantReservation.API/Models/Customers/CustomerDto.cs:1]: every class under models folder can be made Record type

File: RestaurantReservation.API/Models/Customers/CustomerCreationDto.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿namespace RestaurantReservation.API.Models.Customers
+{
+    public class CustomerCreationDto
+    {
+        public string FirstName { get; set; } = string.Empty;
+        public string LastName { get; set; } = string.Empty;
+        public string Email { get; set; }
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Models/Customers/CustomerCreationDto.cs]: this should either have a default value or made nullable

File: RestaurantReservation.API/Profiles/CustomerProfile.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+﻿using AutoMapper;
+using RestaurantReservation.API.Models.Customers;
+using RestaurantReservation.Db.Models.Entities;
+
+namespace RestaurantReservation.API.Profiles
+{
+    public class CustomerProfile : Profile
```
Comment (hiba-kurd) [inline RestaurantReservation.API/Profiles/CustomerProfile.cs:7]: have you looked into other libraries than automapper ?

File: RestaurantReservation.Db/Configurations/OrderConfiguration.cs [hunk 10]
```diff
@@ -22,7 +22,7 @@ public void Configure(EntityTypeBuilder<Order> builder)
             builder.HasOne(r => r.Reservation)
               .WithMany(o => o.Orders)
               .HasForeignKey(r => r.ReservationId)
-              .OnDelete(DeleteBehavior.ClientSetNull);
+              .OnDelete(DeleteBehavior.Cascade);
```
Comment (hiba-kurd) [inline RestaurantReservation.Db/Configurations/OrderConfiguration.cs:25]: what's the reason behind this change ?

File: RestaurantReservation/Program.cs [hunk 11]
```diff
@@ -1,17 +1,33 @@
...
     {
         static async Task Main(string[] args)
         {
-            var context = new RestaurantReservationDbContext();
-            var reservationTester = new ReservationTester(context);
-            await reservationTester.RunTests();
+            var host = Host.CreateDefaultBuilder(args)
+                .ConfigureServices((context, services) =>
+                {
+                    services.AddDbContext<RestaurantReservationDbContext>(options =>
```
Comment (hiba-kurd) [inline RestaurantReservation/Program.cs:17]: you have two Program.cs files 
looks like this project isn't now used with the new ASP project correct ?

File: RestaurantReservation.API/Controllers/CustomersController.cs (hunk 4 above)
Comment (mahamdeh0) [reply RestaurantReservation.API/Controllers/CustomersController.cs:44]: Good question, Adding a service layer between the controller and repository can be a good idea, especially as the application grows,it help keep the controller focused on handling HTTP request while the service layer can take care of any business logic or additional processin before interacting with the database, tthis separation make the code more modular and easier to test since we can unit test the service layer independently, Plus if we need to add more complex logic in the future having a service layer already in place makes it simpler to manage

File: RestaurantReservation.API/Controllers/EmployeesController.cs (hunk 7 above)
Comment (mahamdeh0) [reply RestaurantReservation.API/Controllers/EmployeesController.cs]: i added the trycatch in this endpoint because i faced an issue during test and i wanted to handle potential error specificaly for this case, generally i understand that exception handling should be done at a higher level like in middleware to avoid repeating the same code in every controller method this was just a temporary solution during the testing phase

File: RestaurantReservation.API/Profiles/CustomerProfile.cs (hunk 9 above)
Comment (mahamdeh0) [reply RestaurantReservation.API/Profiles/CustomerProfile.cs:7]: yes ihave worked with AutoMapper
but I’ve also explored other libraries like Mapster this alternative are faster and lighter, and i sometimes prefer them for simpler mapping or when performance is critical
however i ended up using AutoMapper because I was more familiar with it

File: RestaurantReservation/Program.cs (hunk 11 above)
Comment (mahamdeh0) [reply RestaurantReservation/Program.cs:17]: yes that's correct

File: RestaurantReservation.Db/Configurations/OrderConfiguration.cs (hunk 10 above)
Comment (mahamdeh0) [reply RestaurantReservation.Db/Configurations/OrderConfiguration.cs:25]: the change ensures that when a reservation is deleted all related order are also deleted automaticaly this behavior maintains data integrity by preventing orphaned records and ensuring that the system remains consistent as each order should not exist without a valid reservation

File: RestaurantReservation.API/Controllers/CustomersController.cs (hunk 4 above)
Comment (mahamdeh0) [reply RestaurantReservation.API/Controllers/CustomersController.cs:44]: Good question, Adding a service layer between the controller and repository can be a good idea, especially as the application grows,it help keep the controller focused on handling HTTP request while the service layer can take care of any business logic or additional processin before interacting with the database, tthis separation make the code more modular and easier to test since we can unit test the service layer independently, Plus if we need to add more complex logic in the future having a service layer already in place makes it simpler to manage
